docs/changes/shared-plugins-during-build.md

```diff
@@ -15,8 +14,68 @@ Affect scope: `Vite Plugin Authors`
 
 ## Motivation
 
-// TODO:
+Align dev and build plugin pipelines.
 
 ## Migration Guide
 
-// TODO:
+To be able to share plugins across environments, plugin state must be keyed by the current environment. A plugin of the following form will count the number of transformed modules across all environments.
+
+```js
+function CountTransformedModulesPlugin() {
+  let transformedModules
+  return {
+    name: 'count-transformed-modules',
+    buildStart() {
+      transformedModules = 0
+    },
+    transform(id) {
+      transformedModules++
+    },
+    buildEnd() {
+      console.log(transformedModules)
+    },
+  }
+}
+```
+
+If we instead want to count the number of transformed modules for each environment, we need to keep a map:
+
+```js
+function PerEnvironmentCountTransformedModulesPlugin() {
+  const state = new Map<Environment, { count: number }>()
+  return {
+    name: 'count-transformed-modules',
+    perEnvironmentBuildStartEnd: true,
```

Comment:
Does `perEnvironmentBuildStartEnd` still exist? The search result does not return any usage only the declarations: https://github.com/search?q=repo%3Avitejs%2Fvite+perEnvironmentBuildStartEnd&type=code

Reply:
Sorry, it is actually called `perEnvironmentStartEndDuringDev`. We need a better name for this. Any recommendations?

About this option, the problem we have is that plugins my be expecting a single `buildStart` call during dev, even if there are multiple environments, so I don't think we can start calling it multiple times in Vite 6. A lot of times it will work though, because it will end up reseting multiple times the state, but it could fail too. Moving to multiple `buildStart` calls is something we haven't discuss much but I think it is a good idea to align dev and build

Reply:
Ah, that was it. I don't have any better naming 😅. But I guess we should rename `perEnvironmentBuildStartEnd` to `perEnvironmentStartEndDuringDev` and use the settings in the pluginContainer.